File: app/src/main/java/com/callmangement/EHR/imagepicker/ui/imagepicker/ImagePicker.java
/*
 * Created by Nguyen Hoang Lam
 * Date: ${DATE}
 */

package com.callmangement.EHR.imagepicker.ui.imagepicker;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.content.res.Resources;

import androidx.fragment.app.Fragment;

import com.callmangement.R;
import com.callmangement.EHR.imagepicker.model.Config;
import com.callmangement.EHR.imagepicker.model.Image;
import com.callmangement.EHR.imagepicker.model.SavePath;
import com.callmangement.EHR.imagepicker.ui.camera.CameraActivty;

import java.util.ArrayList;

/**
 * Created by hoanglam on 8/4/16.
 */
public class ImagePicker {

    protected Config config;
    public static int RC_PICK_IMAGES_CUSTOM = 1;

    public ImagePicker(Builder builder) {
        config = builder.config;
    }

    public static Builder with(Activity activity) {
        return new ActivityBuilder(activity);
    }

    public static Builder with(Fragment fragment) {
        return new FragmentBuilder(fragment);
    }

    static class ActivityBuilder extends Builder {
        private final Activity activity;

        public ActivityBuilder(Activity activity) {
            super(activity);
            this.activity = activity;
        }

        @Override
        public void start(int resultCode) {
            RC_PICK_IMAGES_CUSTOM = resultCode;
            Intent intent;
            if (!config.isCameraOnly()) {
                intent = new Intent(activity, ImagePickerActivity.class);
                intent.putExtra(Config.EXTRA_CONFIG, config);
                activity.startActivityForResult(intent, RC_PICK_IMAGES_CUSTOM);
            } else {
                intent = new Intent(activity, CameraActivty.class);
                intent.putExtra(Config.EXTRA_CONFIG, config);
                intent.addFlags(Intent.FLAG_ACTIVITY_NO_ANIMATION);
                activity.overridePendingTransition(0, 0);
                activity.startActivityForResult(intent, RC_PICK_IMAGES_CUSTOM);
            }
        }
    }

    static class FragmentBuilder extends Builder {
        private final Fragment fragment;

        public FragmentBuilder(Fragment fragment) {
            super(fragment);
            this.fragment = fragment;
        }

        @Override
        public void start(int resultCode) {
            Intent intent;
            if (!config.isCameraOnly()) {
                intent = new Intent(fragment.getActivity(), ImagePickerActivity.class);
                intent.putExtra(Config.EXTRA_CONFIG, config);
                fragment.startActivityForResult(intent, RC_PICK_IMAGES_CUSTOM);
            } else {
                intent = new Intent(fragment.getActivity(), CameraActivty.class);
                intent.putExtra(Config.EXTRA_CONFIG, config);
                intent.addFlags(Intent.FLAG_ACTIVITY_NO_ANIMATION);
                fragment.getActivity().overridePendingTransition(0, 0);
                fragment.startActivityForResult(intent, RC_PICK_IMAGES_CUSTOM);
            }
        }
    }

    public static abstract class Builder extends BaseBuilder {

        public Builder(Activity activity) {
            super(activity);
        }

        public Builder(Fragment fragment) {
            super(fragment.getContext());
        }

        public Builder setToolbarColor(String toolbarColor) {
            config.setToolbarColor(toolbarColor);
            return this;
        }

        public Builder setStatusBarColor(String statusBarColor) {
            config.setStatusBarColor(statusBarColor);
            return this;
        }

        public Builder setToolbarTextColor(String toolbarTextColor) {
            config.setToolbarTextColor(toolbarTextColor);
            return this;
        }

        public Builder setToolbarIconColor(String toolbarIconColor) {
            config.setToolbarIconColor(toolbarIconColor);
            return this;
        }

        public Builder setProgressBarColor(String progressBarColor) {
            config.setProgressBarColor(progressBarColor);
            return this;
        }

        public Builder setBackgroundColor(String backgroundColor) {
            config.setBackgroundColor(backgroundColor);
            return this;
        }

        public Builder setCameraOnly(boolean isCameraOnly) {
            config.setCameraOnly(isCameraOnly);
            return this;
        }

        public Builder setMultipleMode(boolean isMultipleMode) {
            config.setMultipleMode(isMultipleMode);
            return this;
        }

        public Builder setFolderMode(boolean isFolderMode) {
            config.setFolderMode(isFolderMode);
            return this;
        }

        public Builder setShowCamera(boolean isShowCamera) {
            config.setShowCamera(isShowCamera);
            return this;
        }

        public Builder setMaxSize(int maxSize) {
            config.setMaxSize(maxSize);
            return this;
        }

        public Builder setDoneTitle(String doneTitle) {
            config.setDoneTitle(doneTitle);
            return this;
        }

        public Builder setFolderTitle(String folderTitle) {
            config.setFolderTitle(folderTitle);
            return this;
        }

        public Builder setImageTitle(String imageTitle) {
            config.setImageTitle(imageTitle);
            return this;
        }

        public Builder setSavePath(String path) {
            config.setSavePath(new SavePath(path, false));
            return this;
        }

        public Builder setSelectedImages(ArrayList<Image> selectedImages) {
            config.setSelectedImages(selectedImages);
            return this;
        }

        public abstract void start(int resultCode);

    }

    public static abstract class BaseBuilder {

        protected Config config;

        public BaseBuilder(Context context) {
            this.config = new Config();

            Resources resources = context.getResources();
            config.setCameraOnly(false);
            config.setMultipleMode(true);
            config.setFolderMode(true);
            config.setShowCamera(true);
            config.setMaxSize(Config.MAX_SIZE);
            config.setDoneTitle(resources.getString(R.string.imagepicker_action_done));
            config.setFolderTitle(resources.getString(R.string.imagepicker_title_folder));
            config.setImageTitle(resources.getString(R.string.imagepicker_title_image));
            config.setSavePath(SavePath.DEFAULT);
            config.setSelectedImages(new ArrayList<Image>());
        }
    }

}

File: app/src/main/java/com/callmangement/ehr/imagepicker/ui/imagepicker/ImagePicker.kt
/*
 * Created by Nguyen Hoang Lam
 * Date: ${DATE}
 */
package com.callmangement.ehr.imagepicker.ui.imagepicker

import android.app.Activity
import android.content.Context
import android.content.Intent
import androidx.fragment.app.Fragment
import com.callmangement.ehr.imagepicker.model.Config
import com.callmangement.ehr.imagepicker.model.Image
import com.callmangement.ehr.imagepicker.model.SavePath
import com.callmangement.ehr.imagepicker.ui.camera.CameraActivty
import com.callmangement.R

/**
 * Created by hoanglam on 8/4/16.
 */
class ImagePicker(builder: Builder) {
    protected var config: Config = builder.config

    internal class ActivityBuilder(private val activity: Activity) : Builder(
        activity
    ) {
        override fun start(resultCode: Int) {
            RC_PICK_IMAGES_CUSTOM = resultCode
            val intent: Intent
            if (!config.isCameraOnly) {
                intent = Intent(activity, ImagePickerActivity::class.java)
                intent.putExtra(Config.EXTRA_CONFIG, config)
                activity.startActivityForResult(intent, RC_PICK_IMAGES_CUSTOM)
            } else {
                intent = Intent(activity, CameraActivty::class.java)
                intent.putExtra(Config.EXTRA_CONFIG, config)
                intent.addFlags(Intent.FLAG_ACTIVITY_NO_ANIMATION)
                activity.overridePendingTransition(0, 0)
                activity.startActivityForResult(intent, RC_PICK_IMAGES_CUSTOM)
            }
        }
    }

    internal class FragmentBuilder(private val fragment: Fragment) : Builder(
        fragment
    ) {
        override fun start(resultCode: Int) {
            val intent: Intent
            if (!config.isCameraOnly) {
                intent = Intent(fragment.activity, ImagePickerActivity::class.java)
                intent.putExtra(Config.EXTRA_CONFIG, config)
                fragment.startActivityForResult(intent, RC_PICK_IMAGES_CUSTOM)
            } else {
                intent = Intent(fragment.activity, CameraActivty::class.java)
                intent.putExtra(Config.EXTRA_CONFIG, config)
                intent.addFlags(Intent.FLAG_ACTIVITY_NO_ANIMATION)
                fragment.requireActivity().overridePendingTransition(0, 0)
                fragment.startActivityForResult(intent, RC_PICK_IMAGES_CUSTOM)
            }
        }
    }

    abstract class Builder : BaseBuilder {
        constructor(activity: Activity) : super(activity)

        constructor(fragment: Fragment) : super(fragment.requireContext())

        fun setToolbarColor(toolbarColor: String?): Builder {
            config.setToolbarColor(toolbarColor)
            return this
        }

        fun setStatusBarColor(statusBarColor: String?): Builder {
            config.setStatusBarColor(statusBarColor)
            return this
        }

        fun setToolbarTextColor(toolbarTextColor: String?): Builder {
            config.setToolbarTextColor(toolbarTextColor)
            return this
        }

        fun setToolbarIconColor(toolbarIconColor: String?): Builder {
            config.setToolbarIconColor(toolbarIconColor)
            return this
        }

        fun setProgressBarColor(progressBarColor: String?): Builder {
            config.setProgressBarColor(progressBarColor)
            return this
        }

        fun setBackgroundColor(backgroundColor: String?): Builder {
            config.setBackgroundColor(backgroundColor)
            return this
        }

        fun setCameraOnly(isCameraOnly: Boolean): Builder {
            config.isCameraOnly = isCameraOnly
            return this
        }

        fun setMultipleMode(isMultipleMode: Boolean): Builder {
            config.isMultipleMode = isMultipleMode
            return this
        }

        fun setFolderMode(isFolderMode: Boolean): Builder {
            config.isFolderMode = isFolderMode
            return this
        }

        fun setShowCamera(isShowCamera: Boolean): Builder {
            config.isShowCamera = isShowCamera
            return this
        }

        fun setMaxSize(maxSize: Int): Builder {
            config.maxSize = maxSize
            return this
        }

        fun setDoneTitle(doneTitle: String?): Builder {
            config.doneTitle = doneTitle
            return this
        }

        fun setFolderTitle(folderTitle: String?): Builder {
            config.folderTitle = folderTitle
            return this
        }

        fun setImageTitle(imageTitle: String?): Builder {
            config.imageTitle = imageTitle
            return this
        }

        fun setSavePath(path: String?): Builder {
            config.savePath = SavePath(path, false)
            return this
        }

        fun setSelectedImages(selectedImages: MutableList<Image>?): Builder {
            config.selectedImages = selectedImages
            return this
        }

        abstract fun start(resultCode: Int)
    }

    abstract class BaseBuilder(context: Context) {
        var config: Config =
            Config()

        init {
            val resources = context.resources
            config.isCameraOnly = false
            config.isMultipleMode = true
            config.isFolderMode = true
            config.isShowCamera = true
            config.maxSize = Config.MAX_SIZE
            config.doneTitle = resources.getString(R.string.imagepicker_action_done)
            config.folderTitle = resources.getString(R.string.imagepicker_title_folder)
            config.imageTitle = resources.getString(R.string.imagepicker_title_image)
            config.savePath = SavePath.DEFAULT
            config.selectedImages =
                ArrayList()
        }
    }

    companion object {
        @JvmField
        var RC_PICK_IMAGES_CUSTOM: Int = 1

        @JvmStatic
        fun with(activity: Activity): Builder {
            return ActivityBuilder(activity)
        }

        fun with(fragment: Fragment): Builder {
            return FragmentBuilder(fragment)
        }
    }
}

